carbon-design-system/carbon
# fix(Tabs): allow props to be optional and do not return element array

Closes #14197

#### Changelog

**Changed**

- Set the `leftOverflowButtonProps` and `rightOverflowButtonProps` properties of `TabListProps` to be optional. Both of them are optional in the original propTypes.
- Wrap the return value of `TabPanels` in a fragment based on the following reasons.
  - Returning `undefined` from a component is allowed from React 18. (See https://github.com/reactwg/react-18/discussions/75.)
  - Returning `JSX.Element[]` is allowed from TypeScript v5.1.3. (See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65135.)
  - Returning a fragment can eliminate the dependency on a specific version of React or TypeScript.

#### Testing / Reviewing

- CI should pass.
- Existing Tabs stories should work the same as before.
- Manually verify no TypeScript errors will be detected by following the steps below:
  - Remane `Tabs.stories.js` to `Tabs.stories.tsx`.
  - Add `'../src/**/*.stories.tsx'` to `storyGlobs` in `packages/react/.storybook/main.js`. (See code snippets below.)
  - Enable TypeScript checker in `packages/react/.storybook/main.js`. (See code snippets below.)
  - Start storybook in `packages/react` by running `yarn storybook` and check errors in the terminal. (Yeah there are still some errors, but they are not related to Tabs. For example, some `TextField` components in `Tabs.stories.tsx` do not have the `id` property. Those errors are beyond the scope of this PR.)

```javascript
const storyGlobs =
  process.env.STORYBOOK_STORE_7 === 'false'
    ? [
        '../src/**/*.stories.js',
        '../src/**/next/*.stories.js',
        '../src/**/next/**/*.stories.js',
        '../src/**/*-story.js',
      ]
    : [
        './Welcome/Welcome.mdx',
        '../src/**/*.stories.js',
        '../src/**/*.stories.tsx', // Add this line
        '../src/**/*.stories.mdx',
        '../src/components/Tile/Tile.mdx',
        '../src/**/next/*.stories.js',
        '../src/**/next/**/*.stories.js',
        '../src/**/next/*.stories.mdx',
        '../src/**/*-story.js',
      ];
```

```javascript
const config = {
  // Keep all other fields in this `config` object.
  typescript: {
    reactDocgen: 'react-docgen', // Favor docgen from prop-types instead of TS interfaces
    check: true, // Add this line
  },
```
